File: examples/platform/test_platform_api.py
"""
Test script for the Scienceverse platform API (submit → poll → download).

Mirrors the R sv_convert() function for debugging the async job queue.

Usage:
    python examples/platform/test_platform_api.py <file.pdf>
    python examples/platform/test_platform_api.py <file.pdf> --url http://localhost:8001
    PLATFORM_API_KEY=sv_... python examples/platform/test_platform_api.py paper.pdf
"""

import argparse
import os
import sys
import time

import httpx


def main():
    parser = argparse.ArgumentParser(description="Test Scienceverse platform API")
    parser.add_argument("file", help="Path to document file")
    parser.add_argument(
        "--url",
        default=os.getenv("PLATFORM_API_URL", "https://platform.metacheck.app"),
        help="Platform API base URL",
    )
    parser.add_argument(
        "--api-key",
        default=os.getenv("PLATFORM_API_KEY", ""),
        help="Platform API key (sv_...)",
    )
    parser.add_argument(
        "--poll-interval",
        type=float,
        default=2.0,
        help="Seconds between status polls",
    )
    parser.add_argument(
        "--timeout",
        type=float,
        default=600.0,
        help="Maximum seconds to wait",
    )
    parser.add_argument(
        "--output",
        default=None,
        help="Output path (default: <filename>.json in current dir)",
    )
    parser.add_argument(
        "--format",
        choices=["json"],
        default="json",
        help="Result format to download",
    )
    args = parser.parse_args()

    if not args.api_key:
        print("Error: No API key. Set PLATFORM_API_KEY or pass --api-key.", file=sys.stderr)
        sys.exit(1)

    if not os.path.exists(args.file):
        print(f"Error: File not found: {args.file}", file=sys.stderr)
        sys.exit(1)

    headers = {"Authorization": f"Bearer {args.api_key}"}
    client = httpx.Client(base_url=args.url, headers=headers, timeout=60)

    # ── 1. Health check ──────────────────────────────────────────────────────
    print(f"Platform: {args.url}")
    try:
        health = client.get("/health")
        print(f"Health:   {health.status_code} {health.json()}")
    except Exception as e:
        print(f"Health check failed: {e}", file=sys.stderr)
        sys.exit(1)

    try:
        ready = client.get("/ready")
        print(f"Ready:    {ready.status_code} {ready.json()}")
    except Exception as e:
        print(f"Readiness check failed: {e}")

    # ── 2. Submit job ────────────────────────────────────────────────────────
    filename = os.path.basename(args.file)
    print(f"\nSubmitting {filename} ...")

    with open(args.file, "rb") as f:
        resp = client.post("/jobs", files={"file": (filename, f)})

    if resp.status_code != 200:
        print(f"Submit failed ({resp.status_code}): {resp.text}", file=sys.stderr)
        sys.exit(1)

    job = resp.json()
    job_id = job["job_id"]
    print(f"Job ID:   {job_id}")
    print(f"Status:   {job['status']}")

    # ── 3. Poll for completion ───────────────────────────────────────────────
    print("\nPolling ...")
    elapsed = 0.0
    last_stage = None

    while elapsed < args.timeout:
        time.sleep(args.poll_interval)
        elapsed += args.poll_interval

        status_resp = client.get(f"/jobs/{job_id}")
        status = status_resp.json()

        stage = status.get("stage", "")
        if stage != last_stage:
            print(f"  [{elapsed:6.1f}s] {status['status']}" + (f" ({stage})" if stage else ""))
            last_stage = stage

        if status["status"] == "complete":
            break
        elif status["status"] == "failed":
            print(f"\nJob failed: {status.get('stage', 'unknown error')}", file=sys.stderr)
            sys.exit(1)
    else:
        print(f"\nTimed out after {args.timeout}s (last: {status['status']})", file=sys.stderr)
        sys.exit(1)

    # ── 4. Download result ───────────────────────────────────────────────────
    print(f"\nDownloading result (format={args.format}) ...")

    result_resp = client.get(
        f"/jobs/{job_id}/result",
        params={"format": args.format},
        timeout=120,
    )

    if result_resp.status_code != 200:
        print(f"Download failed ({result_resp.status_code}): {result_resp.text}", file=sys.stderr)
        sys.exit(1)

    content = result_resp.content
    output_path = args.output or os.path.splitext(filename)[0] + ".json"

    with open(output_path, "wb") as f:
        f.write(content)

    print(f"Saved:    {output_path} ({len(content):,} bytes)")

    # ── 5. Quick peek at result ──────────────────────────────────────────────
    import json

    data = json.loads(content)
    info = data.get("info", {})
    print(f"\nTitle:        {info.get('title', 'N/A')}")
    print(f"DOI:          {info.get('doi', 'N/A')}")
    print(f"Authors:      {len(data.get('author', []))}")
    print(f"References:   {len(data.get('bib', []))}")
    print(f"Sections:     {len(data.get('section', []))}")
    print(f"Sentences:    {len(data.get('text', []))}")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: examples/platform/test_platform_api.R
# Test script for the Scienceverse platform API (submit -> poll -> download).
#
# Mirrors the metacheck sv_convert() function for debugging the async job queue.
#
# Usage:
#   Rscript examples/platform/test_platform_api.R <file.pdf>
#   PLATFORM_API_KEY=sv_... Rscript examples/platform/test_platform_api.R paper.pdf

library(httr2)
library(jsonlite)

`%||%` <- function(x, fallback) if (is.null(x)) fallback else x

# ── Configuration ────────────────────────────────────────────────────────────

args <- commandArgs(trailingOnly = TRUE)
if (length(args) == 0) {
  stop("Usage: Rscript test_platform_api.R <file_path> [api_url]")
}

FILE_PATH  <- args[1]
API_URL    <- if (length(args) >= 2) args[2] else "https://platform.metacheck.app"
API_KEY    <- Sys.getenv("PLATFORM_API_KEY")
POLL_INTERVAL <- 2
TIMEOUT    <- 600

if (nchar(API_KEY) == 0) {
  stop("Set the PLATFORM_API_KEY environment variable")
}
if (!file.exists(FILE_PATH)) {
  stop("File not found: ", FILE_PATH)
}

cat("Platform:", API_URL, "\n")
cat("File:    ", basename(FILE_PATH), "\n\n")

# ── 1. Health check ─────────────────────────────────────────────────────────

tryCatch({
  health <- request(API_URL) |>
    req_url_path_append("health") |>
    req_auth_bearer_token(API_KEY) |>
    req_perform() |>
    resp_body_json()
  cat("Health:  ", health$status, "\n")
}, error = \(e) cat("Health check failed:", e$message, "\n"))

tryCatch({
  ready <- request(API_URL) |>
    req_url_path_append("ready") |>
    req_auth_bearer_token(API_KEY) |>
    req_perform() |>
    resp_body_json()
  cat("Ready:   ", ready$status, "\n")
}, error = \(e) cat("Ready check failed:", e$message, "\n"))

# ── 2. Submit job ────────────────────────────────────────────────────────────

cat("\nSubmitting", basename(FILE_PATH), "...\n")

submit_resp <- request(API_URL) |>
  req_url_path_append("jobs") |>
  req_auth_bearer_token(API_KEY) |>
  req_body_multipart(file = curl::form_file(FILE_PATH)) |>
  req_timeout(60) |>
  req_perform()

if (resp_status(submit_resp) != 200) {
  stop("Submit failed (HTTP ", resp_status(submit_resp), "): ",
       resp_body_string(submit_resp))
}

job <- resp_body_json(submit_resp)
job_id <- job$job_id
cat("Job ID:  ", job_id, "\n")
cat("Status:  ", job$status, "\n")

# ── 3. Poll for completion ───────────────────────────────────────────────────

cat("\nPolling ...\n")
elapsed <- 0
last_stage <- ""

repeat {
  Sys.sleep(POLL_INTERVAL)
  elapsed <- elapsed + POLL_INTERVAL

  status_resp <- request(API_URL) |>
    req_url_path_append("jobs", job_id) |>
    req_auth_bearer_token(API_KEY) |>
    req_timeout(30) |>
    req_perform()

  status <- resp_body_json(status_resp)
  stage <- status$stage %||% ""

  if (stage != last_stage) {
    cat(sprintf("  [%5.1fs] %s%s\n",
                elapsed, status$status,
                if (nchar(stage) > 0) paste0(" (", stage, ")") else ""))
    last_stage <- stage
  }

  if (identical(status$status, "complete")) break

  if (identical(status$status, "failed")) {
    stop("Job failed: ", status$stage %||% "unknown error")
  }

  if (elapsed >= TIMEOUT) {
    stop("Timed out after ", TIMEOUT, "s (last: ", status$status, ")")
  }
}

# ── 4. Download result ───────────────────────────────────────────────────────

cat("\nDownloading result ...\n")

result_resp <- request(API_URL) |>
  req_url_path_append("jobs", job_id, "result") |>
  req_auth_bearer_token(API_KEY) |>
  req_timeout(120) |>
  req_perform()

if (resp_status(result_resp) != 200) {
  stop("Download failed (HTTP ", resp_status(result_resp), ")")
}

json_string <- resp_body_string(result_resp)

# Save JSON to disk
json_path <- gsub("\\.[^.]+$", ".json", basename(FILE_PATH))
writeLines(json_string, json_path)
cat("Saved:   ", json_path, " (", nchar(json_string), " chars)\n\n")

# ── 5. Quick peek at result ──────────────────────────────────────────────────

data <- fromJSON(json_string, simplifyVector = TRUE, flatten = TRUE)
metadata <- data$metadata %||% data$info

cat("bibr version:", data$extraction$producer$version %||% data$extraction$bibr_version %||%
  metadata$bibr_version, "\n")
cat("Paper ID:    ", data$paper_id, "\n\n")

# Helper for safe row count
nrow_safe <- function(x) {
  if (is.null(x) || length(x) == 0) return(0L)
  nrow(as.data.frame(x))
}

table_names <- c("author", "text", "section", "url", "bib", "xref", "figure", "table", "eq")
for (tbl_name in table_names) {
  n <- nrow_safe(data[[tbl_name]])
  cat(sprintf("  %-10s %d rows\n", paste0(tbl_name, ":"), n))
}

# Show title if available
if (!is.null(metadata$title)) {
  cat("\nTitle:      ", metadata$title, "\n")
}
if (!is.null(metadata$doi)) {
  cat("DOI:        ", metadata$doi, "\n")
}

cat("\nDone.\n")
